R/check-up-to-date.R

```diff
@@ -0,0 +1,146 @@
+#' Check model and data up to date with outputs
+#'
+#' Functions for checking that the model outputs on disk match the md5 hashes
+#' stored in `bbi_config.json` at run time. In other words, checking that the
+#' model and data files have _not_ changed since the model was last run. See
+#' Details section for what specific files are checked.
+#'
+#' @details
+#' Different files are checked depending on what type of model is being checked.
+#'
+#' **For NONMEM models**
+#'
+#' * The model file (control stream)
+#'
+#' * The data file (referenced in `$DATA` within the control stream)
+#'
+#' **Currently only NONMEM implemented.**
+#'
+#' @param .bbi_object the object to check. Could be
+#'   a `bbi_{.model_type}_model` object,
+#'   a `bbi_{.model_type}_summary` object,
+#'   or a `bbi_log_df` tibble.
+#' @param ... Arguments passed through (currently none).
+#'
+#' @return
+#'
+#' **`bbi_model`** method invisibly returns a logical vector of length 2. The
+#' first element (named `"model"`) refers to the model files mentioned above.
+#' The second element (named `"data"`) refers to the data files mentioned above.
+#' For both elements, they will be `TRUE` if nothing has changed, `FALSE` if
+#' anything has changed. Note: _if no file exists_ at the specified path,
+#' `FALSE` will be returned because that is technically a "change." The file
+#' used to exist and now it does not.
+#'
+#' **`bbi_log_df`** method invisibly returns a named list of lists, with one
+#' element for each row in the input tibble, with the name corresponding to the
+#' value in the `run` column for that row. Each element of the list will contain
+#' the two-element list returned from the `bbi_model` method (described above)
+#' for the relevant model.
+#'
+#' There is no `add_up_to_date()` function because **if you would like to add
+#' these columns to a `bbi_log_df` tibble** you can use [add_config()], which
+#' contains `model_has_changed` and `data_has_changed` columns. Please note:
+#' these contain the opposite boolean values (`check_up_to_date()` returns
+#' `TRUE` if up to date, `*_has_changed` returns `TRUE` if _changed_).
+#'
+#' **The returned value is invisible because a message is printed** alerting the
+#' user of the specific files that have changed, if any. This facilitates
+#' calling the function for this side effect without explicitly handling the
+#' returned value.
+#'
+#' @export
+check_up_to_date <- function(.bbi_object, ...) {
+  UseMethod("check_up_to_date")
+}
+
+#' @export
+check_up_to_date.bbi_nonmem_model <- function(.bbi_object, ...) {
```

Comment:
Part of the point of this is prepare for adding Stan stuff so, even though these are the same, they're _not_ gonna be the default methods. I considered adding another parent class that they would share, but that seemed a little excessive to have something like
```r
class(.mod)
# "bbi_nonmem_model"  "bbi_nonmem_object"  "bbi_model"  "list"
```
(`"bbi_model"` is already the parent class that they'll share with Stan and other kinds of models)

Anyway, that would be a totally valid way to do it, but I opted for just duplicating a couple dispatches instead (and made them use the same private helper to avoid copy/paste).